pkg/reconcilermanager/controllers/reconciler_base.go

```diff
@@ -236,17 +245,95 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
+		if err := deploymentResourceInterface.Delete(ctx, dRef.Name, metav1.DeleteOptions{}); err != nil {
 			return controllerutil.OperationResultNone, err
 		}
-		if err := r.client.Create(ctx, declared); err != nil {
+		data, err := json.Marshal(declared)
+		if err != nil {
+			return controllerutil.OperationResultNone, fmt.Errorf("failed to marshal declared deployment object to byte array: %w", err)
+		}
+		_, err = deploymentResourceInterface.Patch(ctx, dRef.Name, types.ApplyPatchType, data, metav1.PatchOptions{FieldManager: fieldManager, Force: &forcePatch})
+		if err != nil {
 			return controllerutil.OperationResultNone, err
 		}
 	}
-
+	if deploymentAfterPatch.GetGeneration() == unObjCurrent.GetGeneration() {
+		return controllerutil.OperationResultNone, nil
+	}
 	return controllerutil.OperationResultUpdated, nil
 }
 
+// preprocessDeployment delete all the fields in allowlist from unstructured object and convert the unstructured object to Deployment object
+func preprocessDeployment(allowList []string, unstructuredDeployment *unstructured.Unstructured) (*appsv1.Deployment, error) {
+	for _, path := range allowList {
+		if err := deleteFields(unstructuredDeployment.Object, path); err != nil {
+			return nil, err
+		}
+	}
+	var resultDeployment appsv1.Deployment
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredDeployment.Object, &resultDeployment); err != nil {
+		return nil, fmt.Errorf("failed to convert from current reconciler unstructured object to deployment object: %w", err)
+	}
+	return &resultDeployment, nil
+}
+
+type deploymentProcessResult struct {
+	isDeploymentSame     bool
+	isDeploymentAdjusted bool
+	deploymentToPatch    []byte
+}
+
+// isDeploymentSame checks if current deployment is same with declared deployment when ignore the fields in allowlist.
+func isDeploymentSame(isAutopilot bool, declared *appsv1.Deployment, unObjCurrent *unstructured.Unstructured, scheme *runtime.Scheme) (*deploymentProcessResult, error) {
+	//The fields in allowlist are the fields we allow
+	// users or other controllers to modify.
+	allowList := []string{
+		"$.spec.template.spec.containers[*].terminationMessagePath",
+		"$.spec.template.spec.containers[*].imagePullPolicy",
+		"$.spec.template.spec.containers[*].terminationMessagePolicy",
+		"$.spec.template.spec.containers[*].resources.requests.ephemeral-storage",
+		"$.spec.template.spec.containers[*].*.timeoutSeconds",
+		"$.spec.template.spec.containers[*].*.periodSeconds",
+		"$.spec.template.spec.containers[*].*.successThreshold",
+		"$.spec.template.spec.containers[*].*.failureThreshold",
+		"$.spec.template.spec.tolerations",
+		"$.spec.template.spec.restartPolicy",
+		"$.spec.template.spec.nodeSelector",
+		"$.spec.template.spec.terminationGracePeriodSeconds",
+		"$.spec.template.spec.dnsPolicy",
+		"$.spec.template.spec.schedulerName",
+		"$.spec.revisionHistoryLimit",
+		"$.spec.progressDeadlineSeconds",
+		"$.spec.paused",
```

Comment:
Yes, this means we allow users or other controllers to pause the reconciler, and since this field is not defined in the declared deployment, users and other controllers will be able to pause the reconciler. I add this field here is to make sure when that pause set to true by users / other controllers, reconciler manager does not keep patching the reconciler deployment. If we want to make sure others cannot pause reconciler, we should define this field in reconciler-manager-configmap.

Reply:
Removed paused from allowList